npm_and_yarn/lib/dependabot/npm_and_yarn/helpers.rb

```diff
@@ -453,7 +470,47 @@ def self.fallback_to_local_version(name)
         package_manager_activate(name, current_version)
       end
 
-      # Install the package manager for specified version by using corepack
+      # Find cached version that matches the requested version pattern
+      sig { params(name: String, version: String).returns(T.nilable(String)) }
+      def self.find_cached_version(name, version)
+        cache_dir = "/home/dependabot/.cache/node/corepack/v1/#{name}"
+        return nil unless Dir.exist?(cache_dir)
+
+        cached_versions = list_cached_versions(cache_dir)
+
+        # Check for exact version match first
+        return version if cached_versions.include?(version)
+
+        # Handle major version resolution
+        find_highest_major_version(cached_versions, version)
+      rescue StandardError => e
+        Dependabot.logger.debug("Cache detection failed for #{name}@#{version}: #{e.message}")
+        nil
+      end
+
+      # List all cached versions, excluding system entries
+      sig { params(cache_dir: String).returns(T::Array[String]) }
+      def self.list_cached_versions(cache_dir)
+        Dir.entries(cache_dir).reject { |entry| entry.start_with?(".") }
+      end
+
+      # Find the highest cached version for a major version request
+      sig { params(cached_versions: T::Array[String], version: String).returns(T.nilable(String)) }
+      def self.find_highest_major_version(cached_versions, version)
+        # Only optimize for simple major version requests (e.g., "11", not "11.6.2")
+        return nil unless version.match?(/^\d+$/)
+
+        major = version.to_i
+        matching_versions = cached_versions.select { |v| v.match?(/^#{major}\./) }
+
+        return nil if matching_versions.empty?
```

Comment:
Is this safe? What if the user requests `11.6.2` but what we have cached is `11.5.0`?